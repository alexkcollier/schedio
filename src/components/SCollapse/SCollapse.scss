.collapse {
  $root: &;

  &:not(:last-child) {
    border-bottom: border();
    border-color: color('grey', 'lighter');
    margin-bottom: spacing('half') + spacing('quarter');
  }

  &__trigger {
    display: flex;
    justify-content: space-between;
    margin-bottom: spacing('half');
    width: 100%;
  }

  &__icon {
    margin-left: spacing('half');
    transition: all duration() easing();
  }

  &__content {
    padding: spacing();
    padding-top: spacing('half') + spacing('quarter');
  }

  &--active {
    #{$root}__trigger {
      background: color('ice');
    }

    #{$root}__icon {
      transform: rotate(-180deg);
    }
  }
}
