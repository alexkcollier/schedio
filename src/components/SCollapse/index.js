import SCollapse from './SCollapse.vue'

export { SCollapse }

export default {
  install: Vue => Vue.component(SCollapse.name, SCollapse),
}
