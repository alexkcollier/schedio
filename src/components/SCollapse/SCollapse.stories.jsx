import { SCollapse } from '@/components/SCollapse'
import { SHeading } from '@/components/SHeading'
import PropList from '@@/docs/components/PropList'
import { storiesOf } from '@storybook/vue'
import { text } from '@storybook/addon-knobs'

storiesOf('Components/SCollapse', module)
  .addParameters({ jest: 'SCollapse' })
  .add('Collapse', () => {
    return {
      props: {
        props: {
          default: () => ({
            content: text('content', 'Collapse content', 'Optional props'),
            triggerText: text('trigger-text', 'Collapse trigger text', 'Optional props'),
          }),
        },
        slots: {
          default: () => ({
            content: text('v-slot:content', 'Collapse content slot', '$slots'),
            triggerText: text('v-slot:trigger-text', 'Collapse trigger text slot', '$slots'),
          }),
        },
      },
      render (h) {
        const { props, slots } = this.$props

        return (
          <div>
            <SHeading>Collapse</SHeading>
            <p>
              <code>{'<SCollapse />'}</code> can be used to toggle the visibility of content.
            </p>

            <SHeading level="2">Example</SHeading>

            <div>
              <SCollapse {...{ props }} />

              <SCollapse>
                <template slot="trigger-text">{slots.triggerText}</template>
                <div slot="content">{slots.content}</div>
              </SCollapse>

              <SCollapse />
            </div>

            {/* This doesn't work if props are only generated through a mixin */}
            {SCollapse.props && <PropList component={SCollapse} />}
          </div >
        )
      },
    }
  })
