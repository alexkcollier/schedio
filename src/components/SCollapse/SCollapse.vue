<template>
  <article
    class="collapse"
    :class="{
      'collapse--active': showContent,
    }"
  >
    <header>
      <SButton
        class="collapse__trigger"
        @click="toggleContent"
      >
        <slot name="trigger-text">
          {{ triggerText }}
        </slot>

        <SIcon
          class="collapse__icon"
          icon="chevron-down"
          aria-hidden="true"
        />
      </SButton>
    </header>

    <Transition name="fade">
      <main
        v-show="showContent"
        class="collapse__content"
        :aria-hidden="!showContent"
      >
        <slot name="content">
          {{ content }}
        </slot>
      </main>
    </Transition>
  </article>
</template>

<script>
import { SButton } from '@/components/SButton';
import { SIcon } from '@/components/SIcon';

export default {
  name: 'SCollapse',

  components: {
    SButton,
    SIcon,
  },

  props: {
    triggerText: {
      type: String,
      default: 'Collapse trigger text',
    },

    content: {
      type: String,
      default: 'Collapse content',
    },
  },

  data () {
    return {
      showContent: false,
    };
  },

  methods: {
    toggleContent () {
      this.showContent = !this.showContent;
    },
  },
};
</script>

<style lang="scss">
.collapse {
  $root: &;

  &:not(:last-child) {
    border-bottom: border();
    border-color: color('grey', 'lighter');
    margin-bottom: spacing('tight');
  }

  &__trigger {
    display: flex;
    justify-content: space-between;
    margin-bottom: spacing('half');
    text-align: left;
    width: 100%;
  }

  &__icon {
    flex: 0 0 auto;
    margin-left: spacing('half');
    transition: all duration() easing();
  }

  &__content {
    padding: spacing();
    padding-top: spacing('tight');
  }

  &--active {
    #{$root}__trigger {
      background: color('ice');
    }

    #{$root}__icon {
      transform: rotate(-180deg);
    }
  }
}
</style>
