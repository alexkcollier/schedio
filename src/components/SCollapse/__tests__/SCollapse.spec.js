import SCollapse from '@/components/SCollapse/SCollapse.vue'
import { mount, shallowMount } from '@vue/test-utils'
import { axe, toHaveNoViolations } from 'jest-axe'

expect.extend(toHaveNoViolations)

describe('SCollapse.vue', () => {
  const content = 'content'
  const triggerText = 'trigger text'
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(SCollapse)
  })

  it('renders correctly', async () => {
    expect(wrapper.html()).toMatchSnapshot()
    expect(await axe(wrapper.html())).toHaveNoViolations()
  })

  it('opens and closes', () => {
    wrapper = mount(SCollapse)
    const button = wrapper.find('button')

    button.trigger('click')
    expect(wrapper.vm.showContent).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()

    button.trigger('click')
    expect(wrapper.vm.showContent).toBe(false)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('accepts props for text', () => {
    wrapper.setProps({ content, triggerText })
    expect(wrapper.text()).toContain(content)
    expect(wrapper.text()).toContain(triggerText)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('accepts slots for text', () => {
    wrapper = shallowMount(SCollapse, {
      slots: { content, triggerText },
    })

    expect(wrapper.text()).toContain(content)
    expect(wrapper.text()).toContain(triggerText)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
