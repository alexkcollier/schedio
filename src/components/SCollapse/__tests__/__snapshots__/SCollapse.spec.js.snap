// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SCollapse.vue accepts props for text 1`] = `
<article class="collapse">
  <header>
    <sbutton-stub color="" shade="" size="" type="" iconleft="" iconright="" class="collapse__trigger">
      Collapse trigger text
      <sicon-stub icon="chevron-down" color="" shade="" size="" aria-hidden="true" class="collapse__icon"></sicon-stub>
    </sbutton-stub>
  </header>
  <transition-stub name="fade">
    <main aria-hidden="true" class="collapse__content" style="display: none;">
      Collapse content
    </main>
  </transition-stub>
</article>
`;

exports[`SCollapse.vue accepts slots for text 1`] = `
<article class="collapse">
  <header>
    <sbutton-stub color="" shade="" size="" type="" iconleft="" iconright="" class="collapse__trigger">
      Collapse trigger text
      <sicon-stub icon="chevron-down" color="" shade="" size="" aria-hidden="true" class="collapse__icon"></sicon-stub>
    </sbutton-stub>
  </header>
  <transition-stub name="fade">
    <main aria-hidden="true" class="collapse__content" style="display: none;">content</main>
  </transition-stub>
</article>
`;

exports[`SCollapse.vue opens and closes 1`] = `
<article class="collapse">
  <header><button class="button collapse__trigger">
      <!---->
      <!---->
      Collapse trigger text
      <svg viewBox="0 0 24 24" class="icon collapse__icon" aria-hidden="true">
        <polyline points="6 9 12 15 18 9"></polyline>
      </svg>
      <!----></button></header>
  <main aria-hidden="true" class="collapse__content" style="display: none;">
    Collapse content
  </main>
</article>
`;

exports[`SCollapse.vue opens and closes 2`] = `
<article class="collapse">
  <header><button class="button collapse__trigger">
      <!---->
      <!---->
      Collapse trigger text
      <svg viewBox="0 0 24 24" class="icon collapse__icon" aria-hidden="true">
        <polyline points="6 9 12 15 18 9"></polyline>
      </svg>
      <!----></button></header>
  <main aria-hidden="true" class="collapse__content" style="display: none;">
    Collapse content
  </main>
</article>
`;

exports[`SCollapse.vue renders correctly 1`] = `
<article class="collapse">
  <header>
    <sbutton-stub color="" shade="" size="" type="" iconleft="" iconright="" class="collapse__trigger">
      Collapse trigger text
      <sicon-stub icon="chevron-down" color="" shade="" size="" aria-hidden="true" class="collapse__icon"></sicon-stub>
    </sbutton-stub>
  </header>
  <transition-stub name="fade">
    <main aria-hidden="true" class="collapse__content" style="display: none;">
      Collapse content
    </main>
  </transition-stub>
</article>
`;
